src/ucp/core/ucp_am.c

```diff
@@ -508,69 +570,91 @@ ucp_am_send_req(ucp_request_t *req, size_t count,
      */
     status = ucp_request_send(req, 0);
     if (req->flags & UCP_REQUEST_FLAG_COMPLETED) {
-        ucs_trace_req("releasing send request %p, returning status %s", req,
-                      ucs_status_string(status));
-        ucp_request_put(req);
-        return UCS_STATUS_PTR(status);
+        ucp_request_imm_cmpl_param(param, req, status, send);
     }
 
-    ucp_request_set_callback(req, send.cb, (ucp_send_nbx_callback_t)cb, NULL);
+    ucp_request_set_send_callback_param(param, req, send);
 
     return req + 1;
 }
 
-UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_send_nb,
-                 (ep, id, payload, count, datatype, cb, flags),
-                 ucp_ep_h ep, uint16_t id, const void *payload,
-                 size_t count, uintptr_t datatype,
-                 ucp_send_callback_t cb, unsigned flags)
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_am_send_inline(ucp_ep_h ep, uint16_t id, uint32_t flags, const void *header,
+                   size_t header_length, const void *buffer, size_t length)
+{
+    if (ucs_likely(!(flags & UCP_AM_SEND_REPLY) &&
+                   ((length == 0) || (header_length == 0)) &&
```

Comment:
I guess that the `((length == 0) || (header_length == 0))` condition is already covered in
`((ssize_t)(length + header_length) <=
  ucp_ep_config(ep)->am.max_short)`

Reply:
no, user can provide both header and data, which would fit short fragment

Reply:
> no, user can provide both header and data, which would fit short fragment

ah, I see